src/twisted/protocols/_smb/core.py

```diff
@@ -0,0 +1,519 @@
+NEGOTIATE_SIGNING_REQUIRED = 0x0002
+
+MAX_READ_SIZE = 0x10000
+MAX_TRANSACT_SIZE = 0x10000
+MAX_WRITE_SIZE = 0x10000
+
+SESSION_FLAG_IS_GUEST = 0x0001
+SESSION_FLAG_IS_NULL = 0x0002
+SESSION_FLAG_ENCRYPT_DATA = 0x0004
+
+NEGOTIATE_SIGNING_ENABLED = 0x0001
+NEGOTIATE_SIGNING_REQUIRED = 0x0002
+
+GLOBAL_CAP_DFS = 0x00000001
+GLOBAL_CAP_LEASING = 0x00000002
+GLOBAL_CAP_LARGE_MTU = 0x00000004
+GLOBAL_CAP_MULTI_CHANNEL = 0x00000008
+GLOBAL_CAP_PERSISTENT_HANDLES = 0x00000010
+GLOBAL_CAP_DIRECTORY_LEASING = 0x00000020
+GLOBAL_CAP_ENCRYPTION = 0x00000040
+
+MAX_DIALECT = 0x02FF
+
+SHARE_DISK = 0x01
+SHARE_PIPE = 0x02
+SHARE_PRINTER = 0x03
+
+SHAREFLAG_MANUAL_CACHING = 0x00000000
+SHAREFLAG_AUTO_CACHING = 0x00000010
+SHAREFLAG_VDO_CACHING = 0x00000020
+SHAREFLAG_NO_CACHING = 0x00000030
+SHAREFLAG_DFS = 0x00000001
+SHAREFLAG_DFS_ROOT = 0x00000002
+SHAREFLAG_RESTRICT_EXCLUSIVE_OPENS = 0x00000100
+SHAREFLAG_FORCE_SHARED_DELETE = 0x00000200
+SHAREFLAG_ALLOW_NAMESPACE_CACHING = 0x00000400
+SHAREFLAG_ACCESS_BASED_DIRECTORY_ENUM = 0x00000800
+SHAREFLAG_FORCE_LEVELII_OPLOCK = 0x00001000
+SHAREFLAG_ENABLE_HASH_V1 = 0x00002000
+SHAREFLAG_ENABLE_HASH_V2 = 0x00004000
+SHAREFLAG_ENCRYPT_DATA = 0x00008000
+SHAREFLAG_IDENTITY_REMOTING = 0x00040000
+
+SHARE_CAP_DFS = 0x00000008
+SHARE_CAP_CONTINUOUS_AVAILABILITY = 0x00000010
+SHARE_CAP_SCALEOUT = 0x00000020
+SHARE_CAP_CLUSTER = 0x00000040
+SHARE_CAP_ASYMMETRIC = 0x00000080
+SHARE_CAP_REDIRECT_TO_OWNER = 0x00000100
+
+FILE_READ_DATA = 0x00000001
+FILE_LIST_DIRECTORY = 0x00000001
+FILE_WRITE_DATA = 0x00000002
+FILE_ADD_FILE = 0x00000002
+FILE_APPEND_DATA = 0x00000004
+FILE_ADD_SUBDIRECTORY = 0x00000004
+FILE_READ_EA = 0x00000008  # "Extended Attributes"
+FILE_WRITE_EA = 0x00000010
+FILE_DELETE_CHILD = 0x00000040
+FILE_EXECUTE = 0x00000020
+FILE_TRAVERSE = 0x00000020
+FILE_READ_ATTRIBUTES = 0x00000080
+FILE_WRITE_ATTRIBUTES = 0x00000100
+DELETE = 0x00010000
+READ_CONTROL = 0x00020000
+WRITE_DAC = 0x00040000
+WRITE_OWNER = 0x00080000
+SYNCHRONIZE = 0x00100000
+ACCESS_SYSTEM_SECURITY = 0x01000000
+MAXIMUM_ALLOWED = 0x02000000
+GENERIC_ALL = 0x10000000
+GENERIC_EXECUTE = 0x20000000
+GENERIC_WRITE = 0x40000000
+GENERIC_READ = 0x80000000
+
+
+
+class SMBConnection(base.SMBPacketReceiver):
+    """
+    implement SMB protocol server-side
+    """
+    def __init__(self, factory, addr):
+        base.SMBPacketReceiver.__init__(self)
+        log.debug("new SMBConnection from {addr!r}", addr=addr)
+        self.addr = addr
+        self.factory = factory
+        self.avatar = None
+        self.logout_thunk = None
+        self.signing_enabled = False
+        self.signing_required = False
+        self.message_id = 0
+        self.tree_id = 0
+        self.session_id = 0
+        self.async_id = 0
+        self.first_session_setup = True
+        self.is_async = False
+        self.is_related = False
+        self.is_signed = False
+        self.blob_manager = security_blob.BlobManager(factory.domain)
+        self.trees = {}
```

Comment:
Surely we would want all of these attributes to be private?

Reply:
Also, Twisted naming convention would be `isAsync`, `isRelated`, `isSigned`, etc

Reply:
if they are private that assumes SMBConnection will handle every packet type, which would make it huge. It would be nice to farm out some work to a "session" object : but it will need access to these header variables somehow